Improve make_tensor performance for float and complex types

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #85473

For floating types, `make_tensor` calls `rand` and then does a linear
interpolation from `low` to `high`. This instead calls `uniform_(low,
high)` to cut out the interpolation step.

For complex types, `make_tensor` does the `rand` + interpolation step
twice and calls `torch.complex(real, imag)` at the end. This instead
uses `view_as_real` and `uniform_(low, high)` to fuse it all into one
operation.

My benchmarks show significant speedups in all cases for float32 and
complex64.

| Device | dtype     | Size  | Master (us) | This PR (us) | Speedup |
|--------|-----------|-------|-------------|--------------|---------|
| CPU    | float32   | 8     | 19.4        | 6.34         | 3.1     |
|        |           | 4096  | 36.8        | 21.3         | 1.7     |
|        |           | 2**24 | 167,000     | 80,500       | 2.1     |
|        | complex32 | 8     | 37.0        | 7.57         | 4.9     |
|        |           | 4096  | 73.1        | 37.6         | 1.9     |
|        |           | 2**24 | 409,000     | 161,000      | 2.5     |
| CUDA   | float32   | 8     | 40.4        | 11.7         | 3.5     |
|        |           | 4096  | 38.7        | 11.7         | 3.3     |
|        |           | 2**24 | 2,300       | 238          | 9.7     |
|        | complex32 | 8     | 78.7        | 14           | 5.6     |
|        |           | 4096  | 82.7        | 13.8         | 6.0     |
|        |           | 2**24 | 5,520       | 489          | 11.3    |

peterbell10 committed:
make_tensor lerp

[ghstack-poisoned]

peterbell10 committed:
Update on "make_tensor lerp"




[ghstack-poisoned]

peterbell10 committed:
Update on "Improve make_tensor performance for float and complex types"


For floating types, `make_tensor` calls `rand` and then does a linear
interpolation (aka lerp) from `low` to `high`. This makes the lerp
step faster by:

- using inplace operations
- using `add`'s `alpha` parameter to avoid an extra kernel
- adding shortcuts for special values of `low` and `high`

For complex types, `make_tensor` does the `rand` + interpolation step
twice and calls `torch.complex(real, imag)` at the end. This reduce
overhead by doing a single `rand` + interpolate of double the size,
then calling `torch.view_as_complex` at the end.

My benchmarks show speedups in all cases for float32 and complex64.

| Device | dtype     | Size  | low | high | Master (us) | This PR (us) | Speedup |
|--------|-----------|-------|-----|------|-------------|--------------|---------|
| CPU    | float32   | 8     |     |      | 19.4        | 15.1         | 1.3     |
|        |           |       | 0   |      | 19.7        | 9.21         | 2.1     |
|        |           |       | 0   | 1    | 19.7        | 5.94         | 3.3     |
|        |           | 4096  |     |      | 36.8        | 31.3         | 1.2     |
|        |           |       | 0   |      | 37.1        | 24.7         | 1.5     |
|        |           |       | 0   | 1    | 36.9        | 21.0         | 1.8     |
|        |           | 2**24 |     |      | 167,000     | 115,000      | 1.5     |
|        |           |       | 0   |      | 179,000     | 85,200       | 2.1     |
|        |           |       | 0   | 1    | 180,000     | 80,800       | 2.2     |
|        | complex32 | 8     |     |      | 37.0        | 17.6         | 2.1     |
|        |           |       | 0   |      | 37.4        | 11.3         | 3.3     |
|        |           |       | 0   | 1    | 37.5        | 7.66         | 4.9     |
|        |           | 4096  |     |      | 73.1        | 49.9         | 1.5     |
|        |           |       | 0   |      | 73.5        | 41.5         | 1.8     |
|        |           |       | 0   | 1    | 73.6        | 37.6         | 2.0     |
|        |           | 2**24 |     |      | 409,000     | 229,000      | 1.8     |
|        |           |       | 0   |      | 411,000     | 170,000      | 2.4     |
|        |           |       | 0   | 1    | 409,000     | 163,000      | 2.5     |
| CUDA   | float32   | 8     |     |      | 40.4        | 30.9         | 1.3     |
|        |           |       | 0   |      | 39.2        | 17.6         | 2.2     |
|        |           |       | 0   | 1    | 39.2        | 11.1         | 3.5     |
|        |           | 4096  |     |      | 38.7        | 32.2         | 1.2     |
|        |           |       | 0   |      | 39.2        | 18.0         | 2.2     |
|        |           |       | 0   | 1    | 39.3        | 11.1         | 3.5     |
|        |           | 2**24 |     |      | 2,300       | 1,840        | 1.3     |
|        |           |       | 0   |      | 2,300       | 704          | 3.3     |
|        |           |       | 0   | 1    | 2,300       | 242          | 9.5     |
|        | complex32 | 8     |     |      | 78.7        | 34.7         | 2.3     |
|        |           |       | 0   |      | 80.8        | 20.5         | 3.9     |
|        |           |       | 0   | 1    | 83.5        | 13.8         | 6.0     |
|        |           | 4096  |     |      | 82.7        | 34.8         | 2.4     |
|        |           |       | 0   |      | 83.9        | 20.5         | 4.1     |
|        |           |       | 0   | 1    | 81.5        | 13.9         | 5.9     |
|        |           | 2**24 |     |      | 5,520       | 3,670        | 1.5     |
|        |           |       | 0   |      | 5,520       | 1,400        | 3.9     |
|        |           |       | 0   | 1    | 5,520       | 484          | 11.4    |

[ghstack-poisoned]

peterbell10 committed:
Update on "Improve make_tensor performance for float and complex types"


For floating types, `make_tensor` calls `rand` and then does a linear
interpolation (aka lerp) from `low` to `high`. This makes the lerp
step faster by:

- using inplace operations
- using `add`'s `alpha` parameter to avoid an extra kernel
- adding shortcuts for special values of `low` and `high`

For complex types, `make_tensor` does the `rand` + interpolation step
twice and calls `torch.complex(real, imag)` at the end. This reduce
overhead by doing a single `rand` + interpolate of double the size,
then calling `torch.view_as_complex` at the end.

My benchmarks show speedups in all cases for float32 and complex64.

| Device | dtype     | Size  | low | high | Master (us) | This PR (us) | Speedup |
|--------|-----------|-------|-----|------|-------------|--------------|---------|
| CPU    | float32   | 8     |     |      | 19.4        | 15.1         | 1.3     |
|        |           |       | 0   |      | 19.7        | 9.21         | 2.1     |
|        |           |       | 0   | 1    | 19.7        | 5.94         | 3.3     |
|        |           | 4096  |     |      | 36.8        | 31.3         | 1.2     |
|        |           |       | 0   |      | 37.1        | 24.7         | 1.5     |
|        |           |       | 0   | 1    | 36.9        | 21.0         | 1.8     |
|        |           | 2**24 |     |      | 167,000     | 115,000      | 1.5     |
|        |           |       | 0   |      | 179,000     | 85,200       | 2.1     |
|        |           |       | 0   | 1    | 180,000     | 80,800       | 2.2     |
|        | complex32 | 8     |     |      | 37.0        | 17.6         | 2.1     |
|        |           |       | 0   |      | 37.4        | 11.3         | 3.3     |
|        |           |       | 0   | 1    | 37.5        | 7.66         | 4.9     |
|        |           | 4096  |     |      | 73.1        | 49.9         | 1.5     |
|        |           |       | 0   |      | 73.5        | 41.5         | 1.8     |
|        |           |       | 0   | 1    | 73.6        | 37.6         | 2.0     |
|        |           | 2**24 |     |      | 409,000     | 229,000      | 1.8     |
|        |           |       | 0   |      | 411,000     | 170,000      | 2.4     |
|        |           |       | 0   | 1    | 409,000     | 163,000      | 2.5     |
| CUDA   | float32   | 8     |     |      | 40.4        | 30.9         | 1.3     |
|        |           |       | 0   |      | 39.2        | 17.6         | 2.2     |
|        |           |       | 0   | 1    | 39.2        | 11.1         | 3.5     |
|        |           | 4096  |     |      | 38.7        | 32.2         | 1.2     |
|        |           |       | 0   |      | 39.2        | 18.0         | 2.2     |
|        |           |       | 0   | 1    | 39.3        | 11.1         | 3.5     |
|        |           | 2**24 |     |      | 2,300       | 1,840        | 1.3     |
|        |           |       | 0   |      | 2,300       | 704          | 3.3     |
|        |           |       | 0   | 1    | 2,300       | 242          | 9.5     |
|        | complex32 | 8     |     |      | 78.7        | 34.7         | 2.3     |
|        |           |       | 0   |      | 80.8        | 20.5         | 3.9     |
|        |           |       | 0   | 1    | 83.5        | 13.8         | 6.0     |
|        |           | 4096  |     |      | 82.7        | 34.8         | 2.4     |
|        |           |       | 0   |      | 83.9        | 20.5         | 4.1     |
|        |           |       | 0   | 1    | 81.5        | 13.9         | 5.9     |
|        |           | 2**24 |     |      | 5,520       | 3,670        | 1.5     |
|        |           |       | 0   |      | 5,520       | 1,400        | 3.9     |
|        |           |       | 0   | 1    | 5,520       | 484          | 11.4    |

[ghstack-poisoned]

peterbell10 committed:
Update on "Improve make_tensor performance for float and complex types"


For floating types, `make_tensor` calls `rand` and then does a linear
interpolation (aka lerp) from `low` to `high`. This makes the lerp
step faster by:

- using inplace operations
- using `add`'s `alpha` parameter to avoid an extra kernel
- adding shortcuts for special values of `low` and `high`

For complex types, `make_tensor` does the `rand` + interpolation step
twice and calls `torch.complex(real, imag)` at the end. This reduce
overhead by doing a single `rand` + interpolate of double the size,
then calling `torch.view_as_complex` at the end.

My benchmarks show speedups in all cases for float32 and complex64.

| Device | dtype     | Size  | low | high | Master (us) | This PR (us) | Speedup |
|--------|-----------|-------|-----|------|-------------|--------------|---------|
| CPU    | float32   | 8     |     |      | 19.4        | 15.1         | 1.3     |
|        |           |       | 0   |      | 19.7        | 9.21         | 2.1     |
|        |           |       | 0   | 1    | 19.7        | 5.94         | 3.3     |
|        |           | 4096  |     |      | 36.8        | 31.3         | 1.2     |
|        |           |       | 0   |      | 37.1        | 24.7         | 1.5     |
|        |           |       | 0   | 1    | 36.9        | 21.0         | 1.8     |
|        |           | 2**24 |     |      | 167,000     | 115,000      | 1.5     |
|        |           |       | 0   |      | 179,000     | 85,200       | 2.1     |
|        |           |       | 0   | 1    | 180,000     | 80,800       | 2.2     |
|        | complex32 | 8     |     |      | 37.0        | 17.6         | 2.1     |
|        |           |       | 0   |      | 37.4        | 11.3         | 3.3     |
|        |           |       | 0   | 1    | 37.5        | 7.66         | 4.9     |
|        |           | 4096  |     |      | 73.1        | 49.9         | 1.5     |
|        |           |       | 0   |      | 73.5        | 41.5         | 1.8     |
|        |           |       | 0   | 1    | 73.6        | 37.6         | 2.0     |
|        |           | 2**24 |     |      | 409,000     | 229,000      | 1.8     |
|        |           |       | 0   |      | 411,000     | 170,000      | 2.4     |
|        |           |       | 0   | 1    | 409,000     | 163,000      | 2.5     |
| CUDA   | float32   | 8     |     |      | 40.4        | 30.9         | 1.3     |
|        |           |       | 0   |      | 39.2        | 17.6         | 2.2     |
|        |           |       | 0   | 1    | 39.2        | 11.1         | 3.5     |
|        |           | 4096  |     |      | 38.7        | 32.2         | 1.2     |
|        |           |       | 0   |      | 39.2        | 18.0         | 2.2     |
|        |           |       | 0   | 1    | 39.3        | 11.1         | 3.5     |
|        |           | 2**24 |     |      | 2,300       | 1,840        | 1.3     |
|        |           |       | 0   |      | 2,300       | 704          | 3.3     |
|        |           |       | 0   | 1    | 2,300       | 242          | 9.5     |
|        | complex32 | 8     |     |      | 78.7        | 34.7         | 2.3     |
|        |           |       | 0   |      | 80.8        | 20.5         | 3.9     |
|        |           |       | 0   | 1    | 83.5        | 13.8         | 6.0     |
|        |           | 4096  |     |      | 82.7        | 34.8         | 2.4     |
|        |           |       | 0   |      | 83.9        | 20.5         | 4.1     |
|        |           |       | 0   | 1    | 81.5        | 13.9         | 5.9     |
|        |           | 2**24 |     |      | 5,520       | 3,670        | 1.5     |
|        |           |       | 0   |      | 5,520       | 1,400        | 3.9     |
|        |           |       | 0   | 1    | 5,520       | 484          | 11.4    |

[ghstack-poisoned]

peterbell10 committed:
Update on "Improve make_tensor performance for float and complex types"


For floating types, `make_tensor` calls `rand` and then does a linear
interpolation (aka lerp) from `low` to `high`. This makes the lerp
step faster by:

- using inplace operations
- using `add`'s `alpha` parameter to avoid an extra kernel
- adding shortcuts for special values of `low` and `high`

For complex types, `make_tensor` does the `rand` + interpolation step
twice and calls `torch.complex(real, imag)` at the end. This reduce
overhead by doing a single `rand` + interpolate of double the size,
then calling `torch.view_as_complex` at the end.

My benchmarks show speedups in all cases for float32 and complex64.

| Device | dtype     | Size  | low | high | Master (us) | This PR (us) | Speedup |
|--------|-----------|-------|-----|------|-------------|--------------|---------|
| CPU    | float32   | 8     |     |      | 19.4        | 15.1         | 1.3     |
|        |           |       | 0   |      | 19.7        | 9.21         | 2.1     |
|        |           |       | 0   | 1    | 19.7        | 5.94         | 3.3     |
|        |           | 4096  |     |      | 36.8        | 31.3         | 1.2     |
|        |           |       | 0   |      | 37.1        | 24.7         | 1.5     |
|        |           |       | 0   | 1    | 36.9        | 21.0         | 1.8     |
|        |           | 2**24 |     |      | 167,000     | 115,000      | 1.5     |
|        |           |       | 0   |      | 179,000     | 85,200       | 2.1     |
|        |           |       | 0   | 1    | 180,000     | 80,800       | 2.2     |
|        | complex32 | 8     |     |      | 37.0        | 17.6         | 2.1     |
|        |           |       | 0   |      | 37.4        | 11.3         | 3.3     |
|        |           |       | 0   | 1    | 37.5        | 7.66         | 4.9     |
|        |           | 4096  |     |      | 73.1        | 49.9         | 1.5     |
|        |           |       | 0   |      | 73.5        | 41.5         | 1.8     |
|        |           |       | 0   | 1    | 73.6        | 37.6         | 2.0     |
|        |           | 2**24 |     |      | 409,000     | 229,000      | 1.8     |
|        |           |       | 0   |      | 411,000     | 170,000      | 2.4     |
|        |           |       | 0   | 1    | 409,000     | 163,000      | 2.5     |
| CUDA   | float32   | 8     |     |      | 40.4        | 30.9         | 1.3     |
|        |           |       | 0   |      | 39.2        | 17.6         | 2.2     |
|        |           |       | 0   | 1    | 39.2        | 11.1         | 3.5     |
|        |           | 4096  |     |      | 38.7        | 32.2         | 1.2     |
|        |           |       | 0   |      | 39.2        | 18.0         | 2.2     |
|        |           |       | 0   | 1    | 39.3        | 11.1         | 3.5     |
|        |           | 2**24 |     |      | 2,300       | 1,840        | 1.3     |
|        |           |       | 0   |      | 2,300       | 704          | 3.3     |
|        |           |       | 0   | 1    | 2,300       | 242          | 9.5     |
|        | complex32 | 8     |     |      | 78.7        | 34.7         | 2.3     |
|        |           |       | 0   |      | 80.8        | 20.5         | 3.9     |
|        |           |       | 0   | 1    | 83.5        | 13.8         | 6.0     |
|        |           | 4096  |     |      | 82.7        | 34.8         | 2.4     |
|        |           |       | 0   |      | 83.9        | 20.5         | 4.1     |
|        |           |       | 0   | 1    | 81.5        | 13.9         | 5.9     |
|        |           | 2**24 |     |      | 5,520       | 3,670        | 1.5     |
|        |           |       | 0   |      | 5,520       | 1,400        | 3.9     |
|        |           |       | 0   | 1    | 5,520       | 484          | 11.4    |

[ghstack-poisoned]

peterbell10 committed:
Update on "Improve make_tensor performance for float and complex types"


For floating types, `make_tensor` calls `rand` and then does a linear
interpolation from `low` to `high`. This instead calls `uniform_(low,
high)` to cut out the interpolation step.

For complex types, `make_tensor` does the `rand` + interpolation step
twice and calls `torch.complex(real, imag)` at the end. This instead
uses `view_as_real` and `uniform_(low, high)` to fuse it all into one
operation.

My benchmarks show significant speedups in all cases for float32 and
complex64.

| Device | dtype     | Size  | Master (us) | This PR (us) | Speedup |
|--------|-----------|-------|-------------|--------------|---------|
| CPU    | float32   | 8     | 19.4        | 6.34         | 3.1     |
|        |           | 4096  | 36.8        | 21.3         | 1.7     |
|        |           | 2**24 | 167,000     | 80,500       | 2.1     |
|        | complex32 | 8     | 37.0        | 7.57         | 4.9     |
|        |           | 4096  | 73.1        | 37.6         | 1.9     |
|        |           | 2**24 | 409,000     | 161,000      | 2.5     |
| CUDA   | float32   | 8     | 40.4        | 11.7         | 3.5     |
|        |           | 4096  | 38.7        | 11.7         | 3.3     |
|        |           | 2**24 | 2,300       | 238          | 9.7     |
|        | complex32 | 8     | 78.7        | 14           | 5.6     |
|        |           | 4096  | 82.7        | 13.8         | 6.0     |
|        |           | 2**24 | 5,520       | 489          | 11.3    |

[ghstack-poisoned]

peterbell10 committed:
Update on "Improve make_tensor performance for float and complex types"


For floating types, `make_tensor` calls `rand` and then does a linear
interpolation from `low` to `high`. This instead calls `uniform_(low,
high)` to cut out the interpolation step.

For complex types, `make_tensor` does the `rand` + interpolation step
twice and calls `torch.complex(real, imag)` at the end. This instead
uses `view_as_real` and `uniform_(low, high)` to fuse it all into one
operation.

My benchmarks show significant speedups in all cases for float32 and
complex64.

| Device | dtype     | Size  | Master (us) | This PR (us) | Speedup |
|--------|-----------|-------|-------------|--------------|---------|
| CPU    | float32   | 8     | 19.4        | 6.34         | 3.1     |
|        |           | 4096  | 36.8        | 21.3         | 1.7     |
|        |           | 2**24 | 167,000     | 80,500       | 2.1     |
|        | complex32 | 8     | 37.0        | 7.57         | 4.9     |
|        |           | 4096  | 73.1        | 37.6         | 1.9     |
|        |           | 2**24 | 409,000     | 161,000      | 2.5     |
| CUDA   | float32   | 8     | 40.4        | 11.7         | 3.5     |
|        |           | 4096  | 38.7        | 11.7         | 3.3     |
|        |           | 2**24 | 2,300       | 238          | 9.7     |
|        | complex32 | 8     | 78.7        | 14           | 5.6     |
|        |           | 4096  | 82.7        | 13.8         | 6.0     |
|        |           | 2**24 | 5,520       | 489          | 11.3    |

[ghstack-poisoned]

peterbell10 committed:
Update on "Improve make_tensor performance for float and complex types"


For floating types, `make_tensor` calls `rand` and then does a linear
interpolation from `low` to `high`. This instead calls `uniform_(low,
high)` to cut out the interpolation step.

For complex types, `make_tensor` does the `rand` + interpolation step
twice and calls `torch.complex(real, imag)` at the end. This instead
uses `view_as_real` and `uniform_(low, high)` to fuse it all into one
operation.

My benchmarks show significant speedups in all cases for float32 and
complex64.

| Device | dtype     | Size  | Master (us) | This PR (us) | Speedup |
|--------|-----------|-------|-------------|--------------|---------|
| CPU    | float32   | 8     | 19.4        | 6.34         | 3.1     |
|        |           | 4096  | 36.8        | 21.3         | 1.7     |
|        |           | 2**24 | 167,000     | 80,500       | 2.1     |
|        | complex32 | 8     | 37.0        | 7.57         | 4.9     |
|        |           | 4096  | 73.1        | 37.6         | 1.9     |
|        |           | 2**24 | 409,000     | 161,000      | 2.5     |
| CUDA   | float32   | 8     | 40.4        | 11.7         | 3.5     |
|        |           | 4096  | 38.7        | 11.7         | 3.3     |
|        |           | 2**24 | 2,300       | 238          | 9.7     |
|        | complex32 | 8     | 78.7        | 14           | 5.6     |
|        |           | 4096  | 82.7        | 13.8         | 6.0     |
|        |           | 2**24 | 5,520       | 489          | 11.3    |

[ghstack-poisoned]

peterbell10 committed:
Update on "Improve make_tensor performance for float and complex types"


For floating types, `make_tensor` calls `rand` and then does a linear
interpolation from `low` to `high`. This instead calls `uniform_(low,
high)` to cut out the interpolation step.

For complex types, `make_tensor` does the `rand` + interpolation step
twice and calls `torch.complex(real, imag)` at the end. This instead
uses `view_as_real` and `uniform_(low, high)` to fuse it all into one
operation.

My benchmarks show significant speedups in all cases for float32 and
complex64.

| Device | dtype     | Size  | Master (us) | This PR (us) | Speedup |
|--------|-----------|-------|-------------|--------------|---------|
| CPU    | float32   | 8     | 19.4        | 6.34         | 3.1     |
|        |           | 4096  | 36.8        | 21.3         | 1.7     |
|        |           | 2**24 | 167,000     | 80,500       | 2.1     |
|        | complex32 | 8     | 37.0        | 7.57         | 4.9     |
|        |           | 4096  | 73.1        | 37.6         | 1.9     |
|        |           | 2**24 | 409,000     | 161,000      | 2.5     |
| CUDA   | float32   | 8     | 40.4        | 11.7         | 3.5     |
|        |           | 4096  | 38.7        | 11.7         | 3.3     |
|        |           | 2**24 | 2,300       | 238          | 9.7     |
|        | complex32 | 8     | 78.7        | 14           | 5.6     |
|        |           | 4096  | 82.7        | 13.8         | 6.0     |
|        |           | 2**24 | 5,520       | 489          | 11.3    |

[ghstack-poisoned]

peterbell10 committed:
Update on "Improve make_tensor performance for float and complex types"


For floating types, `make_tensor` calls `rand` and then does a linear
interpolation from `low` to `high`. This instead calls `uniform_(low,
high)` to cut out the interpolation step.

For complex types, `make_tensor` does the `rand` + interpolation step
twice and calls `torch.complex(real, imag)` at the end. This instead
uses `view_as_real` and `uniform_(low, high)` to fuse it all into one
operation.

My benchmarks show significant speedups in all cases for float32 and
complex64.

| Device | dtype     | Size  | Master (us) | This PR (us) | Speedup |
|--------|-----------|-------|-------------|--------------|---------|
| CPU    | float32   | 8     | 19.4        | 6.34         | 3.1     |
|        |           | 4096  | 36.8        | 21.3         | 1.7     |
|        |           | 2**24 | 167,000     | 80,500       | 2.1     |
|        | complex32 | 8     | 37.0        | 7.57         | 4.9     |
|        |           | 4096  | 73.1        | 37.6         | 1.9     |
|        |           | 2**24 | 409,000     | 161,000      | 2.5     |
| CUDA   | float32   | 8     | 40.4        | 11.7         | 3.5     |
|        |           | 4096  | 38.7        | 11.7         | 3.3     |
|        |           | 2**24 | 2,300       | 238          | 9.7     |
|        | complex32 | 8     | 78.7        | 14           | 5.6     |
|        |           | 4096  | 82.7        | 13.8         | 6.0     |
|        |           | 2**24 | 5,520       | 489          | 11.3    |

[ghstack-poisoned]

peterbell10 committed:
Update on "Improve make_tensor performance for float and complex types"


For floating types, `make_tensor` calls `rand` and then does a linear
interpolation from `low` to `high`. This instead calls `uniform_(low,
high)` to cut out the interpolation step.

For complex types, `make_tensor` does the `rand` + interpolation step
twice and calls `torch.complex(real, imag)` at the end. This instead
uses `view_as_real` and `uniform_(low, high)` to fuse it all into one
operation.

My benchmarks show significant speedups in all cases for float32 and
complex64.

| Device | dtype     | Size  | Master (us) | This PR (us) | Speedup |
|--------|-----------|-------|-------------|--------------|---------|
| CPU    | float32   | 8     | 19.4        | 6.34         | 3.1     |
|        |           | 4096  | 36.8        | 21.3         | 1.7     |
|        |           | 2**24 | 167,000     | 80,500       | 2.1     |
|        | complex32 | 8     | 37.0        | 7.57         | 4.9     |
|        |           | 4096  | 73.1        | 37.6         | 1.9     |
|        |           | 2**24 | 409,000     | 161,000      | 2.5     |
| CUDA   | float32   | 8     | 40.4        | 11.7         | 3.5     |
|        |           | 4096  | 38.7        | 11.7         | 3.3     |
|        |           | 2**24 | 2,300       | 238          | 9.7     |
|        | complex32 | 8     | 78.7        | 14           | 5.6     |
|        |           | 4096  | 82.7        | 13.8         | 6.0     |
|        |           | 2**24 | 5,520       | 489          | 11.3    |

[ghstack-poisoned]

peterbell10 committed:
Update on "Improve make_tensor performance for float and complex types"


For floating types, `make_tensor` calls `rand` and then does a linear
interpolation from `low` to `high`. This instead calls `uniform_(low,
high)` to cut out the interpolation step.

For complex types, `make_tensor` does the `rand` + interpolation step
twice and calls `torch.complex(real, imag)` at the end. This instead
uses `view_as_real` and `uniform_(low, high)` to fuse it all into one
operation.

My benchmarks show significant speedups in all cases for float32 and
complex64.

| Device | dtype     | Size  | Master (us) | This PR (us) | Speedup |
|--------|-----------|-------|-------------|--------------|---------|
| CPU    | float32   | 8     | 19.4        | 6.34         | 3.1     |
|        |           | 4096  | 36.8        | 21.3         | 1.7     |
|        |           | 2**24 | 167,000     | 80,500       | 2.1     |
|        | complex32 | 8     | 37.0        | 7.57         | 4.9     |
|        |           | 4096  | 73.1        | 37.6         | 1.9     |
|        |           | 2**24 | 409,000     | 161,000      | 2.5     |
| CUDA   | float32   | 8     | 40.4        | 11.7         | 3.5     |
|        |           | 4096  | 38.7        | 11.7         | 3.3     |
|        |           | 2**24 | 2,300       | 238          | 9.7     |
|        | complex32 | 8     | 78.7        | 14           | 5.6     |
|        |           | 4096  | 82.7        | 13.8         | 6.0     |
|        |           | 2**24 | 5,520       | 489          | 11.3    |

[ghstack-poisoned]

peterbell10 committed:
Update on "Improve make_tensor performance for float and complex types"


For floating types, `make_tensor` calls `rand` and then does a linear
interpolation from `low` to `high`. This instead calls `uniform_(low,
high)` to cut out the interpolation step.

For complex types, `make_tensor` does the `rand` + interpolation step
twice and calls `torch.complex(real, imag)` at the end. This instead
uses `view_as_real` and `uniform_(low, high)` to fuse it all into one
operation.

My benchmarks show significant speedups in all cases for float32 and
complex64.

| Device | dtype     | Size  | Master (us) | This PR (us) | Speedup |
|--------|-----------|-------|-------------|--------------|---------|
| CPU    | float32   | 8     | 19.4        | 6.34         | 3.1     |
|        |           | 4096  | 36.8        | 21.3         | 1.7     |
|        |           | 2**24 | 167,000     | 80,500       | 2.1     |
|        | complex32 | 8     | 37.0        | 7.57         | 4.9     |
|        |           | 4096  | 73.1        | 37.6         | 1.9     |
|        |           | 2**24 | 409,000     | 161,000      | 2.5     |
| CUDA   | float32   | 8     | 40.4        | 11.7         | 3.5     |
|        |           | 4096  | 38.7        | 11.7         | 3.3     |
|        |           | 2**24 | 2,300       | 238          | 9.7     |
|        | complex32 | 8     | 78.7        | 14           | 5.6     |
|        |           | 4096  | 82.7        | 13.8         | 6.0     |
|        |           | 2**24 | 5,520       | 489          | 11.3    |

[ghstack-poisoned]

peterbell10 committed:
Update on "Improve make_tensor performance for float and complex types"


For floating types, `make_tensor` calls `rand` and then does a linear
interpolation from `low` to `high`. This instead calls `uniform_(low,
high)` to cut out the interpolation step.

For complex types, `make_tensor` does the `rand` + interpolation step
twice and calls `torch.complex(real, imag)` at the end. This instead
uses `view_as_real` and `uniform_(low, high)` to fuse it all into one
operation.

My benchmarks show significant speedups in all cases for float32 and
complex64.

| Device | dtype     | Size  | Master (us) | This PR (us) | Speedup |
|--------|-----------|-------|-------------|--------------|---------|
| CPU    | float32   | 8     | 19.4        | 6.34         | 3.1     |
|        |           | 4096  | 36.8        | 21.3         | 1.7     |
|        |           | 2**24 | 167,000     | 80,500       | 2.1     |
|        | complex32 | 8     | 37.0        | 7.57         | 4.9     |
|        |           | 4096  | 73.1        | 37.6         | 1.9     |
|        |           | 2**24 | 409,000     | 161,000      | 2.5     |
| CUDA   | float32   | 8     | 40.4        | 11.7         | 3.5     |
|        |           | 4096  | 38.7        | 11.7         | 3.3     |
|        |           | 2**24 | 2,300       | 238          | 9.7     |
|        | complex32 | 8     | 78.7        | 14           | 5.6     |
|        |           | 4096  | 82.7        | 13.8         | 6.0     |
|        |           | 2**24 | 5,520       | 489          | 11.3    |

[ghstack-poisoned]

peterbell10 committed:
Update on "Improve make_tensor performance for float and complex types"


For floating types, `make_tensor` calls `rand` and then does a linear
interpolation from `low` to `high`. This instead calls `uniform_(low,
high)` to cut out the interpolation step.

For complex types, `make_tensor` does the `rand` + interpolation step
twice and calls `torch.complex(real, imag)` at the end. This instead
uses `view_as_real` and `uniform_(low, high)` to fuse it all into one
operation.

My benchmarks show significant speedups in all cases for float32 and
complex64.

| Device | dtype     | Size  | Master (us) | This PR (us) | Speedup |
|--------|-----------|-------|-------------|--------------|---------|
| CPU    | float32   | 8     | 19.4        | 6.34         | 3.1     |
|        |           | 4096  | 36.8        | 21.3         | 1.7     |
|        |           | 2**24 | 167,000     | 80,500       | 2.1     |
|        | complex32 | 8     | 37.0        | 7.57         | 4.9     |
|        |           | 4096  | 73.1        | 37.6         | 1.9     |
|        |           | 2**24 | 409,000     | 161,000      | 2.5     |
| CUDA   | float32   | 8     | 40.4        | 11.7         | 3.5     |
|        |           | 4096  | 38.7        | 11.7         | 3.3     |
|        |           | 2**24 | 2,300       | 238          | 9.7     |
|        | complex32 | 8     | 78.7        | 14           | 5.6     |
|        |           | 4096  | 82.7        | 13.8         | 6.0     |
|        |           | 2**24 | 5,520       | 489          | 11.3    |

[ghstack-poisoned]

peterbell10 committed:
Update on "Improve make_tensor performance for float and complex types"


For floating types, `make_tensor` calls `rand` and then does a linear
interpolation from `low` to `high`. This instead calls `uniform_(low,
high)` to cut out the interpolation step.

For complex types, `make_tensor` does the `rand` + interpolation step
twice and calls `torch.complex(real, imag)` at the end. This instead
uses `view_as_real` and `uniform_(low, high)` to fuse it all into one
operation.

My benchmarks show significant speedups in all cases for float32 and
complex64.

| Device | dtype     | Size  | Master (us) | This PR (us) | Speedup |
|--------|-----------|-------|-------------|--------------|---------|
| CPU    | float32   | 8     | 19.4        | 6.34         | 3.1     |
|        |           | 4096  | 36.8        | 21.3         | 1.7     |
|        |           | 2**24 | 167,000     | 80,500       | 2.1     |
|        | complex32 | 8     | 37.0        | 7.57         | 4.9     |
|        |           | 4096  | 73.1        | 37.6         | 1.9     |
|        |           | 2**24 | 409,000     | 161,000      | 2.5     |
| CUDA   | float32   | 8     | 40.4        | 11.7         | 3.5     |
|        |           | 4096  | 38.7        | 11.7         | 3.3     |
|        |           | 2**24 | 2,300       | 238          | 9.7     |
|        | complex32 | 8     | 78.7        | 14           | 5.6     |
|        |           | 4096  | 82.7        | 13.8         | 6.0     |
|        |           | 2**24 | 5,520       | 489          | 11.3    |

[ghstack-poisoned]

peterbell10 committed:
Update on "Improve make_tensor performance for float and complex types"


For floating types, `make_tensor` calls `rand` and then does a linear
interpolation from `low` to `high`. This instead calls `uniform_(low,
high)` to cut out the interpolation step.

For complex types, `make_tensor` does the `rand` + interpolation step
twice and calls `torch.complex(real, imag)` at the end. This instead
uses `view_as_real` and `uniform_(low, high)` to fuse it all into one
operation.

My benchmarks show significant speedups in all cases for float32 and
complex64.

| Device | dtype     | Size  | Master (us) | This PR (us) | Speedup |
|--------|-----------|-------|-------------|--------------|---------|
| CPU    | float32   | 8     | 19.4        | 6.34         | 3.1     |
|        |           | 4096  | 36.8        | 21.3         | 1.7     |
|        |           | 2**24 | 167,000     | 80,500       | 2.1     |
|        | complex32 | 8     | 37.0        | 7.57         | 4.9     |
|        |           | 4096  | 73.1        | 37.6         | 1.9     |
|        |           | 2**24 | 409,000     | 161,000      | 2.5     |
| CUDA   | float32   | 8     | 40.4        | 11.7         | 3.5     |
|        |           | 4096  | 38.7        | 11.7         | 3.3     |
|        |           | 2**24 | 2,300       | 238          | 9.7     |
|        | complex32 | 8     | 78.7        | 14           | 5.6     |
|        |           | 4096  | 82.7        | 13.8         | 6.0     |
|        |           | 2**24 | 5,520       | 489          | 11.3    |

[ghstack-poisoned]

peterbell10 committed:
Update on "Improve make_tensor performance for float and complex types"


For floating types, `make_tensor` calls `rand` and then does a linear
interpolation from `low` to `high`. This instead calls `uniform_(low,
high)` to cut out the interpolation step.

For complex types, `make_tensor` does the `rand` + interpolation step
twice and calls `torch.complex(real, imag)` at the end. This instead
uses `view_as_real` and `uniform_(low, high)` to fuse it all into one
operation.

My benchmarks show significant speedups in all cases for float32 and
complex64.

| Device | dtype     | Size  | Master (us) | This PR (us) | Speedup |
|--------|-----------|-------|-------------|--------------|---------|
| CPU    | float32   | 8     | 19.4        | 6.34         | 3.1     |
|        |           | 4096  | 36.8        | 21.3         | 1.7     |
|        |           | 2**24 | 167,000     | 80,500       | 2.1     |
|        | complex32 | 8     | 37.0        | 7.57         | 4.9     |
|        |           | 4096  | 73.1        | 37.6         | 1.9     |
|        |           | 2**24 | 409,000     | 161,000      | 2.5     |
| CUDA   | float32   | 8     | 40.4        | 11.7         | 3.5     |
|        |           | 4096  | 38.7        | 11.7         | 3.3     |
|        |           | 2**24 | 2,300       | 238          | 9.7     |
|        | complex32 | 8     | 78.7        | 14           | 5.6     |
|        |           | 4096  | 82.7        | 13.8         | 6.0     |
|        |           | 2**24 | 5,520       | 489          | 11.3    |

[ghstack-poisoned]

peterbell10 committed:
Update on "Improve make_tensor performance for float and complex types"


For floating types, `make_tensor` calls `rand` and then does a linear
interpolation from `low` to `high`. This instead calls `uniform_(low,
high)` to cut out the interpolation step.

For complex types, `make_tensor` does the `rand` + interpolation step
twice and calls `torch.complex(real, imag)` at the end. This instead
uses `view_as_real` and `uniform_(low, high)` to fuse it all into one
operation.

My benchmarks show significant speedups in all cases for float32 and
complex64.

| Device | dtype     | Size  | Master (us) | This PR (us) | Speedup |
|--------|-----------|-------|-------------|--------------|---------|
| CPU    | float32   | 8     | 19.4        | 6.34         | 3.1     |
|        |           | 4096  | 36.8        | 21.3         | 1.7     |
|        |           | 2**24 | 167,000     | 80,500       | 2.1     |
|        | complex32 | 8     | 37.0        | 7.57         | 4.9     |
|        |           | 4096  | 73.1        | 37.6         | 1.9     |
|        |           | 2**24 | 409,000     | 161,000      | 2.5     |
| CUDA   | float32   | 8     | 40.4        | 11.7         | 3.5     |
|        |           | 4096  | 38.7        | 11.7         | 3.3     |
|        |           | 2**24 | 2,300       | 238          | 9.7     |
|        | complex32 | 8     | 78.7        | 14           | 5.6     |
|        |           | 4096  | 82.7        | 13.8         | 6.0     |
|        |           | 2**24 | 5,520       | 489          | 11.3    |

[ghstack-poisoned]

peterbell10 committed:
Update on "Improve make_tensor performance for float and complex types"


For floating types, `make_tensor` calls `rand` and then does a linear
interpolation from `low` to `high`. This instead calls `uniform_(low,
high)` to cut out the interpolation step.

For complex types, `make_tensor` does the `rand` + interpolation step
twice and calls `torch.complex(real, imag)` at the end. This instead
uses `view_as_real` and `uniform_(low, high)` to fuse it all into one
operation.

My benchmarks show significant speedups in all cases for float32 and
complex64.

| Device | dtype     | Size  | Master (us) | This PR (us) | Speedup |
|--------|-----------|-------|-------------|--------------|---------|
| CPU    | float32   | 8     | 19.4        | 6.34         | 3.1     |
|        |           | 4096  | 36.8        | 21.3         | 1.7     |
|        |           | 2**24 | 167,000     | 80,500       | 2.1     |
|        | complex32 | 8     | 37.0        | 7.57         | 4.9     |
|        |           | 4096  | 73.1        | 37.6         | 1.9     |
|        |           | 2**24 | 409,000     | 161,000      | 2.5     |
| CUDA   | float32   | 8     | 40.4        | 11.7         | 3.5     |
|        |           | 4096  | 38.7        | 11.7         | 3.3     |
|        |           | 2**24 | 2,300       | 238          | 9.7     |
|        | complex32 | 8     | 78.7        | 14           | 5.6     |
|        |           | 4096  | 82.7        | 13.8         | 6.0     |
|        |           | 2**24 | 5,520       | 489          | 11.3    |

[ghstack-poisoned]

peterbell10 committed:
Update on "Improve make_tensor performance for float and complex types"


For floating types, `make_tensor` calls `rand` and then does a linear
interpolation from `low` to `high`. This instead calls `uniform_(low,
high)` to cut out the interpolation step.

For complex types, `make_tensor` does the `rand` + interpolation step
twice and calls `torch.complex(real, imag)` at the end. This instead
uses `view_as_real` and `uniform_(low, high)` to fuse it all into one
operation.

My benchmarks show significant speedups in all cases for float32 and
complex64.

| Device | dtype     | Size  | Master (us) | This PR (us) | Speedup |
|--------|-----------|-------|-------------|--------------|---------|
| CPU    | float32   | 8     | 19.4        | 6.34         | 3.1     |
|        |           | 4096  | 36.8        | 21.3         | 1.7     |
|        |           | 2**24 | 167,000     | 80,500       | 2.1     |
|        | complex32 | 8     | 37.0        | 7.57         | 4.9     |
|        |           | 4096  | 73.1        | 37.6         | 1.9     |
|        |           | 2**24 | 409,000     | 161,000      | 2.5     |
| CUDA   | float32   | 8     | 40.4        | 11.7         | 3.5     |
|        |           | 4096  | 38.7        | 11.7         | 3.3     |
|        |           | 2**24 | 2,300       | 238          | 9.7     |
|        | complex32 | 8     | 78.7        | 14           | 5.6     |
|        |           | 4096  | 82.7        | 13.8         | 6.0     |
|        |           | 2**24 | 5,520       | 489          | 11.3    |

[ghstack-poisoned]

peterbell10 committed:
Update on "Improve make_tensor performance for float and complex types"


For floating types, `make_tensor` calls `rand` and then does a linear
interpolation from `low` to `high`. This instead calls `uniform_(low,
high)` to cut out the interpolation step.

For complex types, `make_tensor` does the `rand` + interpolation step
twice and calls `torch.complex(real, imag)` at the end. This instead
uses `view_as_real` and `uniform_(low, high)` to fuse it all into one
operation.

My benchmarks show significant speedups in all cases for float32 and
complex64.

| Device | dtype     | Size  | Master (us) | This PR (us) | Speedup |
|--------|-----------|-------|-------------|--------------|---------|
| CPU    | float32   | 8     | 19.4        | 6.34         | 3.1     |
|        |           | 4096  | 36.8        | 21.3         | 1.7     |
|        |           | 2**24 | 167,000     | 80,500       | 2.1     |
|        | complex32 | 8     | 37.0        | 7.57         | 4.9     |
|        |           | 4096  | 73.1        | 37.6         | 1.9     |
|        |           | 2**24 | 409,000     | 161,000      | 2.5     |
| CUDA   | float32   | 8     | 40.4        | 11.7         | 3.5     |
|        |           | 4096  | 38.7        | 11.7         | 3.3     |
|        |           | 2**24 | 2,300       | 238          | 9.7     |
|        | complex32 | 8     | 78.7        | 14           | 5.6     |
|        |           | 4096  | 82.7        | 13.8         | 6.0     |
|        |           | 2**24 | 5,520       | 489          | 11.3    |

[ghstack-poisoned]

peterbell10 committed:
Update on "Improve make_tensor performance for float and complex types"


For floating types, `make_tensor` calls `rand` and then does a linear
interpolation from `low` to `high`. This instead calls `uniform_(low,
high)` to cut out the interpolation step.

For complex types, `make_tensor` does the `rand` + interpolation step
twice and calls `torch.complex(real, imag)` at the end. This instead
uses `view_as_real` and `uniform_(low, high)` to fuse it all into one
operation.

My benchmarks show significant speedups in all cases for float32 and
complex64.

| Device | dtype     | Size  | Master (us) | This PR (us) | Speedup |
|--------|-----------|-------|-------------|--------------|---------|
| CPU    | float32   | 8     | 19.4        | 6.34         | 3.1     |
|        |           | 4096  | 36.8        | 21.3         | 1.7     |
|        |           | 2**24 | 167,000     | 80,500       | 2.1     |
|        | complex32 | 8     | 37.0        | 7.57         | 4.9     |
|        |           | 4096  | 73.1        | 37.6         | 1.9     |
|        |           | 2**24 | 409,000     | 161,000      | 2.5     |
| CUDA   | float32   | 8     | 40.4        | 11.7         | 3.5     |
|        |           | 4096  | 38.7        | 11.7         | 3.3     |
|        |           | 2**24 | 2,300       | 238          | 9.7     |
|        | complex32 | 8     | 78.7        | 14           | 5.6     |
|        |           | 4096  | 82.7        | 13.8         | 6.0     |
|        |           | 2**24 | 5,520       | 489          | 11.3    |

[ghstack-poisoned]

peterbell10 committed:
Update on "Improve make_tensor performance for float and complex types"


For floating types, `make_tensor` calls `rand` and then does a linear
interpolation from `low` to `high`. This instead calls `uniform_(low,
high)` to cut out the interpolation step.

For complex types, `make_tensor` does the `rand` + interpolation step
twice and calls `torch.complex(real, imag)` at the end. This instead
uses `view_as_real` and `uniform_(low, high)` to fuse it all into one
operation.

My benchmarks show significant speedups in all cases for float32 and
complex64.

| Device | dtype     | Size  | Master (us) | This PR (us) | Speedup |
|--------|-----------|-------|-------------|--------------|---------|
| CPU    | float32   | 8     | 19.4        | 6.34         | 3.1     |
|        |           | 4096  | 36.8        | 21.3         | 1.7     |
|        |           | 2**24 | 167,000     | 80,500       | 2.1     |
|        | complex32 | 8     | 37.0        | 7.57         | 4.9     |
|        |           | 4096  | 73.1        | 37.6         | 1.9     |
|        |           | 2**24 | 409,000     | 161,000      | 2.5     |
| CUDA   | float32   | 8     | 40.4        | 11.7         | 3.5     |
|        |           | 4096  | 38.7        | 11.7         | 3.3     |
|        |           | 2**24 | 2,300       | 238          | 9.7     |
|        | complex32 | 8     | 78.7        | 14           | 5.6     |
|        |           | 4096  | 82.7        | 13.8         | 6.0     |
|        |           | 2**24 | 5,520       | 489          | 11.3    |

[ghstack-poisoned]

peterbell10 committed:
Update on "Improve make_tensor performance for float and complex types"


For floating types, `make_tensor` calls `rand` and then does a linear
interpolation from `low` to `high`. This instead calls `uniform_(low,
high)` to cut out the interpolation step.

For complex types, `make_tensor` does the `rand` + interpolation step
twice and calls `torch.complex(real, imag)` at the end. This instead
uses `view_as_real` and `uniform_(low, high)` to fuse it all into one
operation.

My benchmarks show significant speedups in all cases for float32 and
complex64.

| Device | dtype     | Size  | Master (us) | This PR (us) | Speedup |
|--------|-----------|-------|-------------|--------------|---------|
| CPU    | float32   | 8     | 19.4        | 6.34         | 3.1     |
|        |           | 4096  | 36.8        | 21.3         | 1.7     |
|        |           | 2**24 | 167,000     | 80,500       | 2.1     |
|        | complex32 | 8     | 37.0        | 7.57         | 4.9     |
|        |           | 4096  | 73.1        | 37.6         | 1.9     |
|        |           | 2**24 | 409,000     | 161,000      | 2.5     |
| CUDA   | float32   | 8     | 40.4        | 11.7         | 3.5     |
|        |           | 4096  | 38.7        | 11.7         | 3.3     |
|        |           | 2**24 | 2,300       | 238          | 9.7     |
|        | complex32 | 8     | 78.7        | 14           | 5.6     |
|        |           | 4096  | 82.7        | 13.8         | 6.0     |
|        |           | 2**24 | 5,520       | 489          | 11.3    |

[ghstack-poisoned]

peterbell10 committed:
Update on "Improve make_tensor performance for float and complex types"


For floating types, `make_tensor` calls `rand` and then does a linear
interpolation from `low` to `high`. This instead calls `uniform_(low,
high)` to cut out the interpolation step.

For complex types, `make_tensor` does the `rand` + interpolation step
twice and calls `torch.complex(real, imag)` at the end. This instead
uses `view_as_real` and `uniform_(low, high)` to fuse it all into one
operation.

My benchmarks show significant speedups in all cases for float32 and
complex64.

| Device | dtype     | Size  | Master (us) | This PR (us) | Speedup |
|--------|-----------|-------|-------------|--------------|---------|
| CPU    | float32   | 8     | 19.4        | 6.34         | 3.1     |
|        |           | 4096  | 36.8        | 21.3         | 1.7     |
|        |           | 2**24 | 167,000     | 80,500       | 2.1     |
|        | complex32 | 8     | 37.0        | 7.57         | 4.9     |
|        |           | 4096  | 73.1        | 37.6         | 1.9     |
|        |           | 2**24 | 409,000     | 161,000      | 2.5     |
| CUDA   | float32   | 8     | 40.4        | 11.7         | 3.5     |
|        |           | 4096  | 38.7        | 11.7         | 3.3     |
|        |           | 2**24 | 2,300       | 238          | 9.7     |
|        | complex32 | 8     | 78.7        | 14           | 5.6     |
|        |           | 4096  | 82.7        | 13.8         | 6.0     |
|        |           | 2**24 | 5,520       | 489          | 11.3    |

[ghstack-poisoned]

mruberry commented on torch/testing/_creation.py:
```diff
@@ -13,6 +13,16 @@
+    # uniform_ requires to-from <= std::numeric_limits<scalar_t>::max()
+    # Work around this by scaling the range before and after the PRNG
+    if high - low >= torch.finfo(t.dtype).max:
+        return t.uniform_(low / 2, high / 2).mul_(2)
```
Should this error if the reduced range is still too large?

peterbell10 replied:
I don't think it's possible for this to fail. However, if I'm wrong then `uniform_` will error directly.

The reason I don't think it can fail is that `low` and `high` are clamped to between `finfo(dtype).min` and `finfo(dtype).max`, so I think the largest possible value for `high/2 - low/2` is exactly `finfo(dtype).max`.

peterbell10 committed:
Update on "Improve make_tensor performance for float and complex types"


For floating types, `make_tensor` calls `rand` and then does a linear
interpolation from `low` to `high`. This instead calls `uniform_(low,
high)` to cut out the interpolation step.

For complex types, `make_tensor` does the `rand` + interpolation step
twice and calls `torch.complex(real, imag)` at the end. This instead
uses `view_as_real` and `uniform_(low, high)` to fuse it all into one
operation.

My benchmarks show significant speedups in all cases for float32 and
complex64.

| Device | dtype     | Size  | Master (us) | This PR (us) | Speedup |
|--------|-----------|-------|-------------|--------------|---------|
| CPU    | float32   | 8     | 19.4        | 6.34         | 3.1     |
|        |           | 4096  | 36.8        | 21.3         | 1.7     |
|        |           | 2**24 | 167,000     | 80,500       | 2.1     |
|        | complex32 | 8     | 37.0        | 7.57         | 4.9     |
|        |           | 4096  | 73.1        | 37.6         | 1.9     |
|        |           | 2**24 | 409,000     | 161,000      | 2.5     |
| CUDA   | float32   | 8     | 40.4        | 11.7         | 3.5     |
|        |           | 4096  | 38.7        | 11.7         | 3.3     |
|        |           | 2**24 | 2,300       | 238          | 9.7     |
|        | complex32 | 8     | 78.7        | 14           | 5.6     |
|        |           | 4096  | 82.7        | 13.8         | 6.0     |
|        |           | 2**24 | 5,520       | 489          | 11.3    |

[ghstack-poisoned]

mruberry reviewed (approved):
Cool! Thanks @peterbell10!

pmeier reviewed (commented):
Thanks @peterbell10! I remember we looked at `torch.distributions.uniform.Uniform` instead of `torch.rand` with scaling, but found that it does the same:

https://github.com/pytorch/pytorch/blob/72b32f164415e9f2e86767b2373c939f8f343d1b/torch/distributions/uniform.py#L75-L76

Could you explain where the changes to the tolerances for the tests and `OpInfo`'s come from? They shouldn't be affected by the change in `make_tensor`, are they?

peterbell10 commented:
@pytorchbot merge -l

huydhn commented:
@pytorchbot revert -m 'Sorry for revert your PR, but it seems to cause a bunch of flaky test in pull an periodic' -c nosignal

Here are some related failures.  They looks flaky:
* Periodic
  * https://hud.pytorch.org/pytorch/pytorch/commit/9fb72ca4941edb37c7529b3750617dc4bb6b4fc1
  * https://hud.pytorch.org/pytorch/pytorch/commit/5f26df0345bef35de9cbf585ca0c1af1cd91b9c8
* Pull
  * https://hud.pytorch.org/pytorch/pytorch/commit/44eefb1376b5a05568f047afde4193e951293625
  * https://hud.pytorch.org/pytorch/pytorch/commit/8fb470e81ad29d82204235194b809b8072942c70
  * https://hud.pytorch.org/pytorch/pytorch/commit/3cdf621fe5c8f8378fda209b8a143443d33b2086

peterbell10 committed:
Update on "Improve make_tensor performance for float and complex types"


For floating types, `make_tensor` calls `rand` and then does a linear
interpolation from `low` to `high`. This instead calls `uniform_(low,
high)` to cut out the interpolation step.

For complex types, `make_tensor` does the `rand` + interpolation step
twice and calls `torch.complex(real, imag)` at the end. This instead
uses `view_as_real` and `uniform_(low, high)` to fuse it all into one
operation.

My benchmarks show significant speedups in all cases for float32 and
complex64.

| Device | dtype     | Size  | Master (us) | This PR (us) | Speedup |
|--------|-----------|-------|-------------|--------------|---------|
| CPU    | float32   | 8     | 19.4        | 6.34         | 3.1     |
|        |           | 4096  | 36.8        | 21.3         | 1.7     |
|        |           | 2**24 | 167,000     | 80,500       | 2.1     |
|        | complex32 | 8     | 37.0        | 7.57         | 4.9     |
|        |           | 4096  | 73.1        | 37.6         | 1.9     |
|        |           | 2**24 | 409,000     | 161,000      | 2.5     |
| CUDA   | float32   | 8     | 40.4        | 11.7         | 3.5     |
|        |           | 4096  | 38.7        | 11.7         | 3.3     |
|        |           | 2**24 | 2,300       | 238          | 9.7     |
|        | complex32 | 8     | 78.7        | 14           | 5.6     |
|        |           | 4096  | 82.7        | 13.8         | 6.0     |
|        |           | 2**24 | 5,520       | 489          | 11.3    |

[ghstack-poisoned]

peterbell10 committed:
Update on "Improve make_tensor performance for float and complex types"


For floating types, `make_tensor` calls `rand` and then does a linear
interpolation from `low` to `high`. This instead calls `uniform_(low,
high)` to cut out the interpolation step.

For complex types, `make_tensor` does the `rand` + interpolation step
twice and calls `torch.complex(real, imag)` at the end. This instead
uses `view_as_real` and `uniform_(low, high)` to fuse it all into one
operation.

My benchmarks show significant speedups in all cases for float32 and
complex64.

| Device | dtype     | Size  | Master (us) | This PR (us) | Speedup |
|--------|-----------|-------|-------------|--------------|---------|
| CPU    | float32   | 8     | 19.4        | 6.34         | 3.1     |
|        |           | 4096  | 36.8        | 21.3         | 1.7     |
|        |           | 2**24 | 167,000     | 80,500       | 2.1     |
|        | complex32 | 8     | 37.0        | 7.57         | 4.9     |
|        |           | 4096  | 73.1        | 37.6         | 1.9     |
|        |           | 2**24 | 409,000     | 161,000      | 2.5     |
| CUDA   | float32   | 8     | 40.4        | 11.7         | 3.5     |
|        |           | 4096  | 38.7        | 11.7         | 3.3     |
|        |           | 2**24 | 2,300       | 238          | 9.7     |
|        | complex32 | 8     | 78.7        | 14           | 5.6     |
|        |           | 4096  | 82.7        | 13.8         | 6.0     |
|        |           | 2**24 | 5,520       | 489          | 11.3    |

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge
